### PR TITLE
Add full submap scan matching in 3D.

### PR DESCRIPTION
**Pull request by wohe**

### CODE REVIEW

File: cartographer/mapping_3d/scan_matching/fast_correlative_scan_matcher.cc [hunk 1]
```diff
@@ -95,10 +95,8 @@ FastCorrelativeScanMatcher::FastCorrelativeScanMatcher(
     const proto::FastCorrelativeScanMatcherOptions& options)
     : options_(options),
       resolution_(hybrid_grid.resolution()),
-      linear_xy_window_size_(
-          common::RoundToInt(options_.linear_xy_search_window() / resolution_)),
-      linear_z_window_size_(
-          common::RoundToInt(options_.linear_z_search_window() / resolution_)),
+      origin_(hybrid_grid.origin()),
+      diameter_in_voxels_(hybrid_grid.grid_size()),
```
Comment (damonkohler) [inline cartographer/mapping_3d/scan_matching/fast_correlative_scan_matcher.cc]: width

File: cartographer/mapping_3d/scan_matching/fast_correlative_scan_matcher.cc [hunk 2]
```diff
@@ -110,19 +108,54 @@ bool FastCorrelativeScanMatcher::Match(
...
+  return MatchWithSearchParameters(search_parameters, initial_pose_estimate,
+                                   coarse_point_cloud, fine_point_cloud,
+                                   min_score, score, pose_estimate);
+}
+
+bool FastCorrelativeScanMatcher::MatchFullSubmap(
+    const Eigen::Quaterniond& gravity_alignment,
+    const sensor::PointCloud& coarse_point_cloud,
+    const sensor::PointCloud& fine_point_cloud, const float min_score,
+    float* score, transform::Rigid3d* pose_estimate) const {
```
Comment (damonkohler) [inline cartographer/mapping_3d/scan_matching/fast_correlative_scan_matcher.cc]: const score, const pose_estimate

File: cartographer/mapping_3d/scan_matching/fast_correlative_scan_matcher.cc [hunk 3]
```diff
@@ -110,19 +108,54 @@ bool FastCorrelativeScanMatcher::Match(
...
+  const transform::Rigid3d initial_pose_estimate(origin_.cast<double>(),
+                                                 gravity_alignment);
+  float point_cloud_radius = 0.f;
```
Comment (damonkohler) [inline cartographer/mapping_3d/scan_matching/fast_correlative_scan_matcher.cc]: width?

File: cartographer/mapping_3d/scan_matching/fast_correlative_scan_matcher.cc [hunk 4]
```diff
@@ -110,19 +108,54 @@ bool FastCorrelativeScanMatcher::Match(
...
+  for (const Eigen::Vector3f& point : coarse_point_cloud) {
+    point_cloud_radius = std::max(point_cloud_radius, point.norm());
+  }
+  const int linear_window_size =
+      diameter_in_voxels_ / 2 + std::ceil(point_cloud_radius);
```
Comment (damonkohler) [inline cartographer/mapping_3d/scan_matching/fast_correlative_scan_matcher.cc]: Off by one voxel?

File: cartographer/mapping_3d/scan_matching/fast_correlative_scan_matcher.h [hunk 5]
```diff
@@ -89,28 +89,55 @@ class FastCorrelativeScanMatcher {
...
+  bool MatchFullSubmap(const Eigen::Quaterniond& gravity_alignment,
+                       const sensor::PointCloud& coarse_point_cloud,
+                       const sensor::PointCloud& fine_point_cloud,
+                       float min_score, float* score,
+                       transform::Rigid3d* pose_estimate) const;
+
  private:
-  DiscreteScan DiscretizeScan(const sensor::PointCloud& point_cloud,
+  struct SearchParameters {
+    const int linear_xy_window_size;     // in voxels
```
Comment (damonkohler) [inline cartographer/mapping_3d/scan_matching/fast_correlative_scan_matcher.h]: Consistency.

Just "voxels" and "radians" and unaligned IIRC

Comment (wohe) [inline cartographer/mapping_3d/scan_matching/fast_correlative_scan_matcher.h]: Done. Formatting: clang-format

**Pull request merged**